### PR TITLE
add account_move_import

### PR DESCRIPTION
**Pull request by luc-demeyer**

### CODE REVIEW

File: account_move_import/wizard/import_move_line_wizard.py
Comment (gdgellatly) [inline account_move_import/wizard/import_move_line_wizard.py]: Hi,

You could replace most of this code using a csv.DictReader class.

File: account_move_import/wizard/import_move_line_wizard.py
Comment (gdgellatly) [inline account_move_import/wizard/import_move_line_wizard.py]: Generally these should be encapsulated in a with statement.

e.g.

``` python
with StringIO.StringIO(lines) as lines:
```

As a side point, generally when using StringIO an import like

``` python
try:
    import cStringIO as StringIO
except ImportError:
    import StringIO
```

and for code like this with potentially a lot of lines the extra performance might count.  Besides every OpenERP installation probably has cStringIO

**Comment by gdgellatly** [comment]
Hi,

Thanks for the module.  Just taking a cursory look, I think it could be massively simplified and a lot more performant, maybe 20% of the current size just by using the standard python csv tools and dictionaries.  Is there a reason you decided not to use DictReader?

Also lots of little things, like the fields.selection having 2 choices.  It really shouldn't matter for delimiter, especially using csv.  Just take whatever delimiter and deal with it, lots of different systems use different delimiter like |, \t or even the field seperator character.

Other small notes
Open except clauses
Model naming is generally accepted now as CapWords.
Most new OCA code uses self.pool[] rather than self.pool.get() syntax

**Comment by luc-demeyer** [comment]
Graeme, 
Thanks for the review. As a feedback on your comments: most of the code in this module dates from V5 (and I started my first modules in V5 with ZERO prior python experience). We published it as a module as from V6 and I only did some PEP8 restyling work before testing it on V8 and submitting to the OCA. All our customer running accounting on Odoo are using it, we haven't seen performance problems yet (even not when uploading thousands of entries in an account move), hence I wanted to share the work with the OCA.
If there are some points in your (very valid) comments that you consider blocking for merge, than I'll fix those.

**Comment by gdgellatly** [comment]
Hi Luc,

Don't worry I have thousands of lines of code from similar background, given I started with ZERO anything experience.

I'm going to abstain from this one as I don't know the particular protocols of this repo, but for me given the effort we have put into our quality lately, and how relatively quick it is for this module I'd prefer to see it updated to a more elegant implementation.  Whether that is with this merge or what I don't know, I'll leave that to the maintainers and how they manage it.

**Comment by charbeljc** [comment]
Hi Luc,

From a QA point of view, the first blocker is that your pull request decreases code coverage by 13%. Given the simplicity of this module, it should be easy to achieve a full code coverage for your module with a very small set of simple unit tests.

This would augment your chances to see your module accepted for merge, and gives you some security to refactor your code around csv.DictReader, as proposed by Graeme.

Using csv.Sniffer to determine the exact dialect from your data would allow you to get rid of the 'csv_separator' field and simplify the user experience with an added robustness (computers usually are better than us at guessing data formats).

Being quite new to Odoo and OCA, I can't promise you that it would sufficient to see your pull request merged, but anyway, your module should be in better shape after these steps.

My concern in the long run for inclusion in OCA would be maybe duplication of efforts. Could you explain me the rationale for not reusing the built-in Odoo csv import/export facilities ?

Regards, 
Charbel

**Comment by luc-demeyer** [comment]
Sometimes it would be better to have functional review/understanding of a module before focusing on a technical code coverage thing, especially for a module like this that we have deployed at many customers over the past 5 years. 
None of our accounting customers (and with accounting I mean FULL accounting, not just invoicing & payments) prefers the completely unusable standard import compared to the ease of use offered by an Excel sheet that you can adapt (with formula’s), than save as csv and import into a journal entry (think about complex salary booking, accruals, cost reassignments, year-end postings, …).

To be clear, I don’t disagree with Graeme’s suggestions, but we have more urgent priorities than fixing working modules which are appreciated very much by our current install base.

**Comment by jgrandguillaume** [comment]
Hi Luc,

Thanks a lot for this contribution, indeed, using the standard import feature to load accounting entries isn't really convenient for the end user. I like this module from a functional point of view "we always have to import accounting entries and everyone is doing his own stuff in this area. Glad to see one of those modules land in OCA.

Now, to be honest, from a code perspective, we can't accept it like this. I perfectly respect that you made it with no knowledge of Python or Odoo in the old time. I also have my share of burden ;) But it is never too late to do better right ?

On my side, what is blocking here is:
- No test at all
- A function (aml_import) of more than 320 lines

I agree this module isn't that complicated, but please, to be accepted here you'll need to at least split the aml_import function and provide a simple test. That's not more than a few hours work and I think it deserve it. 

Thanks again for pushing your work here.

Regards,

Joël

**Comment by luc-demeyer** [comment]
@jgrandguillaume 
Joel, the two small things you mentioned shouldn't take too long.. 
I'll do it.

File: account_move_import/wizard/import_move_line_wizard.py
Comment (charbeljc) [inline account_move_import/wizard/import_move_line_wizard.py]: @luc-demeyer Using csv.Sniffer would eliminate this error case. Users can make mistakes, Python not.
This would simplifiy header processing logic and make it easer to extract as a method later.

Regards

**Comment by luc-demeyer** [comment]
@gdgellatly @charbeljc 
Can you review & test the new code.
I have included your suggestions (the .csv.sniffer suggestion appeared to be a bit of a challenge since failing when using real world csv files but I fixed it via try/exect with fallback to standard csv dialect). The end user who can still change settings. I also added Code Page support.
The code has also been completely redesigned to allow customization via inheritance.

File: account_move_import/wizard/import_move_line_wizard.py
Comment (pedrobaeza) [inline account_move_import/wizard/import_move_line_wizard.py]: This should be put as comment, not as string

**Comment by pedrobaeza** [comment]
@jgrandguillaume @gdgellatly can you please make the final review to this? I have requested to change the name to `account_move_line_import`

File: account_move_line_import/__openerp__.py
Comment (pedrobaeza) [inline account_move_line_import/__openerp__.py]: Short header like https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

File: account_move_line_import/__openerp__.py
Comment (pedrobaeza) [inline account_move_line_import/__openerp__.py]: Include `Odoo Community Association (OCA)`

File: account_move_line_import/models/account_move.py
Comment (pedrobaeza) [inline account_move_line_import/models/account_move.py]: Short header

File: account_move_line_import/tests/test_import_move_line.py
Comment (pedrobaeza) [inline account_move_line_import/tests/test_import_move_line.py]: Short header

File: account_move_line_import/wizard/import_move_line_wizard.py
Comment (pedrobaeza) [inline account_move_line_import/wizard/import_move_line_wizard.py]: Short header

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Thanks for the changes. Some comments

**Comment by gdgellatly** [comment]
@luc-demeyer @pedrobaeza I'm sorry I missed the email, and I'm sorry I do not have an 8.0 instance to do a test.  On a purely static technical review 👍 from me.  What follows is a non-blocking comment that I would like to make.

The method of method dispatch is a little clumsy and could be massively shortened, in future it would be nice if we could just dynamically dispatch it. e.g.

``` python
try:
    func = getattr(self, 'handle_%s' % field_name, getattr(self, 'handle_orm_' % field_type))
except AttributeError:
    raise
```

something like that, not fully looked it through, just wanted to note.

**Comment by luc-demeyer** [comment]
@gdgellatly 
Thanks Graeme. I included your suggestion.

**Comment by luc-demeyer** [comment]
@pedrobaeza 
Pedro,
your changes and those of graeme are included.
PR is (finally) green.
Can you merge ?

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**
